Feature/api guide python client (#4023)
influxdata/docs-v2: content/influxdb/v2.2/api-guide/iot-starter/python.md

- **InfluxDB API**: InfluxDB v2 API.
- **IoT device**: Virtual or physical devices write IoT data to the InfluxDB API.
- **IoT Center UI**: User interface sends requests to IoT Center server and renders views for the browser.
- **IoT Center server**: Server and API receives requests from the UI, sends requests to InfluxDB,
  and processes responses from InfluxDB.

### Create IoT Center 

Use [Flask](https://flask.palletsprojects.com/), a light-weight Python web 
framework, and [Jinja](https://jinja.palletsprojects.com/), a templating engine,
to create your sample IoT Center application. 

1. Create and navigate into a new project directory.
    ```bash
    $ mkdir iotproject && cd $_
    ```
2. Create and activate a Python virtual environment for the new project.
    ```bash
    # Create a new virtual environment named "virtualenv"
    # Python 3.8+
    $ python -m venv virtualenv
    # Activate the virtualenv (OS X & Linux)
    $ source virtualenv/bin/activate
    ```
3. After activation completes, use the `pip` package installer (included with Python) to
install Flask and Jinja. 
    ```bash
    $ pip install Flask
    $ pip install Jinja

>     ```

Reviewer: ~~~suggestion
1. Create and navigate into a new project directory.

    ```bash
    $ mkdir iotproject && cd $_
    ```

2. Create and activate a Python virtual environment for the new project.

    ```bash
    # Create a new virtual environment named "virtualenv"
    # Python 3.8+
    $ python -m venv virtualenv
    # Activate the virtualenv (OS X & Linux)
    $ source virtualenv/bin/activate
    ```

3. After activation completes, use the `pip` package installer (included with Python) to
install Flask and Jinja. 

    ```bash
    $ pip install Flask
    $ pip install Jinja
    ```
~~~

   
### Install influxdb-client Package

> Use pip to install the influxdb-client package in your virtual environment.

Reviewer: ```suggestion
Use `pip` to install the influxdb-client package in your virtual environment.
```

For more information about the package, see [the Python InfluxDB client package](https://pypi.org/project/influxdb-client/).

> ```bash

Reviewer: ~~~suggestion
For more information about the package, see [the Python InfluxDB client package](https://pypi.org/project/influxdb-client/).

```bash
~~~

$ pip install influxdb-client
```

### Create a Flask Application

> In your project, create a `app.py` file that:
> 1. Imports the Flask package.

Reviewer: ```suggestion
In your project, create a `app.py` file that:

1. Imports the Flask package.
```

            }
```
{{% caption %}}[@influxdata/iot-python-app ./api/sensor.py](https://github.com/influxdata/iot-python-api/blob/main/api/sensor.py){{% /caption %}}
{{% /truncate %}}

The Sensor object's function `generate_measurement()` will be used to simulate 
weather data.

## Store Virtual IoT Device to InfluxDB

In this section, you will learn how to use the python client library to store
virtual device information in InfluxDB.

In your `./api` directory, create a `devices.py` file. This file holds the 
core functionality for your app.

Import all necessary classes from `influxdb_client`.

```python
import configparser
from datetime import datetime
from uuid import uuid4
from influxdb_client import Authorization, InfluxDBClient, Permission, PermissionResource, Point, WriteOptions
from influxdb_client.client.authorizations_api import AuthorizationsApi
from influxdb_client.client.bucket_api import BucketsApi
from influxdb_client.client.query_api import QueryApi
from influxdb_client.client.write_api import SYNCHRONOUS

from api.sensor import Sensor
```

Ingest the configuartion key value pairs.

```python
config = configparser.ConfigParser()
config.read('config.ini')
```

> Create and store the device: 
> 1. In `devices.py` create a new function `create_device()`. 

Reviewer: ```suggestion
Create and store the device: 

1. In `devices.py`, create a new function `create_device()`. 
```

   This function should take a `device_id` and store virtual device data to the 
   data bucket.

> 2. Initialize `InfluxDBClient` with the config. Pass in url, token and org to
>    create a connection to the InfluxDB instance.

Reviewer: ```suggestion
2. Initialize `InfluxDBClient` with the config. Pass in your InfluxDB URL, token, and organization to
   create a connection to the InfluxDB instance.
```

> 3. Initialize the `WriteAPI` instance. The `WriteApi` instance writes records 
>    to a specified bucket.

Reviewer 2: ```suggestion
3. Initialize a `WriteAPI` instance for writing data to an InfluxDB bucket.
```

8. Check for any failures then return the `device_id` if the request
was successful. Note that `write_api` returns `None` on success.

```python
def create_device(device_id=None):
    influxdb_client = InfluxDBClient(url=config.get('APP', 'INFLUX_URL'),
                                     token=config.get('APP', 'INFLUX_TOKEN'),
                                     org=config.get('APP', 'INFLUX_ORG'))

    if device_id is None:
        device_id = str(uuid4())

    write_api = influxdb_client.write_api(write_options=SYNCHRONOUS)

    point = Point('deviceauth') \
        .tag("deviceId", device_id) \
        .field('key', f'fake_auth_id_{device_id}') \
        .field('token', f'fake_auth_token_{device_id}')

    client_response = write_api.write(bucket=config.get('APP', 'INFLUX_BUCKET_AUTH'), record=point)

    # write() returns None on success
    if client_response is None:
        return device_id

    # Return None on failure
    return None
```
{{% caption %}}[@influxdata/iot-python-app ./api/devices.py create_device() line 59](https://github.com/influxdata/iot-python-api/blob/main/api/devices.py#L59){{% /caption %}}

## Query Bucket For Device

> After storing device information in an InfluxDB bucket, use the Python client library to query 

Reviewer 2: ```suggestion
After storing device information in an InfluxDB bucket, use the client library to query 
```

Reviewer 2: Trying to keep the text as language-agnostic as possible so we can boilerplate.

records generated by the query.

```python
def get_device(device_id) -> {}:
    influxdb_client = InfluxDBClient(url=config.get('APP', 'INFLUX_URL'),
                                     token=config.get('APP', 'INFLUX_TOKEN'),
                                     org=config.get('APP', 'INFLUX_ORG'))

    # Queries must be formatted with single and double quotes correctly
    query_api = QueryApi(influxdb_client)
    device_id = str(device_id)
    device_filter = f'r.deviceId == "{device_id}" and r._field != "token"'
    flux_query = f'from(bucket: "{config.get("APP", "INFLUX_BUCKET_AUTH")}") ' \
                 f'|> range(start: 0) ' \
                 f'|> filter(fn: (r) => r._measurement == "deviceauth" and {device_filter}) ' \
                 f'|> last()'
    devices = {}

    response = query_api.query(flux_query)

   results = []
       for table in response:
           for record in table.records:
               results.append((record.get_field(), record.get_value()))
    return results
```
{{% caption %}}[@influxdata/iot-python-app ./api/devices.py get_device() line 31](https://github.com/influxdata/iot-python-api/blob/main/api/devices.py#L31){{% /caption %}}

> Create a `QueryApi` instance that queries records from devices bucket using Flux.

Reviewer 2: ```suggestion
Create a `QueryApi` instance and a Flux query that retrieves devices from 'INFLUX_BUCKET_AUTH'.
```

> Create a query template that:
> 1. Retrieves the device bucket ID to query from the application configuration.

Reviewer: ```suggestion
Create a query template that:

1. Retrieves the device bucket ID to query from the application configuration.
```

> 1. Retrieves the device bucket ID to query from the application configuration.

Reviewer 2: ```suggestion
1. Retrieves the 'INFLUX_BUCKET_AUTH' bucket ID from `config`.
```

2. Queries the `deviceauth` measurement.
3. Filters points by a specified device ID.
4. Filters out points with the `token` field.

```python
device_filter = f'r.deviceId == "{device_id}" and r._field != "token"'
flux_query = f'from(bucket: "{config.get("APP", "INFLUX_BUCKET_AUTH")}") ' \
           f'|> range(start: 0) ' \
           f'|> filter(fn: (r) => r._measurement == "deviceauth" and {device_filter}) ' \
           f'|> last()'
```

The code below shows how `FluxTable` is parsed and returned as a list of dicts
for each record returned.

```python
    response = query_api.query(flux_query)
    results = []
    for table in response:
        results.append(table.records[0].values)

    return results
```

## Write Telemetry Data
After device information has been stored, generate telemetry data and write the

> records into InfluxDB.

Reviewer 2: Generate simulated telemetry data for devices and write the data to InfluxDB.

> 1. In `devices.py` create a new function `write_measurements()`. 

Reviewer 2: ```suggestion
1. In `devices.py`, create a `write_measurements()` function 
```

>    This function should take a `device_id` and write simulated weather telemetry 

Reviewer 2: ```suggestion
   that takes a `device_id` and writes simulated weather telemetry 
```

    virtual_device = Sensor()
    coord = virtual_device.geo()

    point = Point("environment") \
        .tag("device", device_id) \
        .tag("TemperatureSensor", "virtual_bme280") \
        .tag("HumiditySensor", "virtual_bme280") \
        .tag("PressureSensor", "virtual_bme280") \
        .field("Temperature", virtual_device.generate_measurement()) \
        .field("Humidity", virtual_device.generate_measurement()) \
        .field("Pressure", virtual_device.generate_measurement()) \
        .field("Lat", coord['latitude']) \
        .field("Lon", coord['latitude']) \
        .time(datetime.utcnow())

    print(f"Writing: {point.to_line_protocol()}")
    client_response = write_api.write(bucket=config.get('APP', 'INFLUX_BUCKET'), record=point)

    # write() returns None on success
    if client_response is None:
        # TODO Maybe also return the data that was written
        return device_id

    # Return None on failure
    return None
```
{{% caption %}}[@influxdata/iot-python-app ./api/devices.py write_measurement() line 84](https://github.com/influxdata/iot-python-api/blob/main/api/devices.py#L84){{% /caption %}}

## Query Telemetry Data

> After telemetry data is written into your bucket, query InfluxDB to retrieve the 
> telemetry data.  

Reviewer 2: ```suggestion
After you write telemetry data to the bucket, query InfluxDB to retrieve the data.
```

> Within `devices.py` create a new function called `get_measurements()`. 

Reviewer 2: ```suggestion
In `devices.py`, create a new function called `get_measurements()`. 
```

> This function should:

Reviewer 2: ```suggestion
This function should do the following:
```

> This function should:
>  1. Take a `device_id` and query for simulated weather telemetry data produced by the virtual device.

Reviewer: ```suggestion
This function should:

 1. Take a `device_id` and query for simulated weather telemetry data produced by the virtual device.
```

>  1. Take a `device_id` and query for simulated weather telemetry data produced by the virtual device.

Reviewer 2: ```suggestion
 1. Take a `device_id` and query for weather telemetry data generated by the virtual device.
```

```python
def get_measurements(device_id):
    influxdb_client = InfluxDBClient(url=config.get('APP', 'INFLUX_URL'),
                                     token=config.get('APP', 'INFLUX_TOKEN'),
                                     org=config.get('APP', 'INFLUX_ORG'))

    # Queries must be formatted with single and double quotes correctly
    query_api = QueryApi(influxdb_client)
    device_id = str(device_id)
    device_filter = f'r.device == "{device_id}"'
    flux_query = f'from(bucket: "{config.get("APP", "INFLUX_BUCKET")}") ' \
                 f'|> range(start: 0) ' \
                 f'|> filter(fn: (r) => r._measurement == "environment" and {device_filter}) ' \
                 f'|> last()'

    response = query_api.query(flux_query)

    # iterate through the result(s)
    results = []
    for table in response:
        results.append(table.records[0].values)

    return results
```
{{% caption %}}[@influxdata/iot-python-app ./api/devices.py get_measurements() line 116](https://github.com/influxdata/iot-python-api/blob/main/api/devices.py#L116){{% /caption %}}

The code below shows how `FluxTable` is parsed and returned as a list of dicts
for each record returned.

> ```python

Reviewer: ~~~suggestion
for each record returned.

```python
~~~

# iterate through the result(s)
    results = []
    for table in response:
        results.append(table.records[0].values)

    return results
```
```python
[
   {
      'result': '_result', 
       'table': 0, 
       '_start': datetime.datetime(1970, 1, 1, 0, 0, tzinfo=tzutc()), 
       '_stop': datetime.datetime(2022, 5, 8, 22, 25, 10, 111697, tzinfo=tzutc()), 
       '_time': datetime.datetime(2022, 5, 5, 17, 25, 48, 57014, tzinfo=tzutc()), 
       '_value': 33.780799865722656, 
       'HumiditySensor': 'virtual_bme280', 
       'PressureSensor': 'virtual_bme280', 
       'TemperatureSensor': 'virtual_bme280', 
       '_field': 'Lat', 
       '_measurement': 'environment', 
       'device': 'test_device_508472435243'
    }
]
```

## Connect the UI to the API

> Create a UI that utilizes the application to query and visualize data from InfluxDB.

Reviewer 2: ```suggestion
Create a UI that uses your application API to query and visualize data from InfluxDB.
```

> Your IoT Dashboard has four main pages.
> - Query Devices

Reviewer: ```suggestion
Your IoT Dashboard has four main pages.

- Query Devices
```

- `write.html`: [Writes telemetry data.](https://github.com/influxdata/iot-python-api/blob/main/templates/write.html)

After creating those files are created, update `base.html` and `app.py` to connect the routes.
`app.py` serves the routes and runs the core logic necessary to populate the UI.

For example, in the `data` route, the app retrieves `device_id_input` from 
`data.html` when user input is submitted. It then calls `get_measurements()` 
with the provided input. After the query runs, `data.html` displays query results.

```python
@app.route('/data', methods=['GET', 'POST'])
def data():
    if request.method == 'GET':
        return render_template('data.html', data=None)
    else:
        device_id = request.form.get('device_id_input', None)
        results = devices.get_measurements(device_id)
        return render_template('data.html', data=results)
```
{{% caption %}}[@influxdata/iot-python-app app.py line 83](https://github.com/influxdata/iot-python-api/blob/main/app.py#L83){{% /caption %}}

After these files are created and updated, run the app to view the completed 
IoT Center at `http://localhost:5000`.

> ```bash

Reviewer: ~~~suggestion
IoT Center at `http://localhost:5000`.

```bash
~~~

    return results
```
{{% caption %}}[@influxdata/iot-python-app ./api/devices.py get_measurements() line 116](https://github.com/influxdata/iot-python-api/blob/main/api/devices.py#L116){{% /caption %}}

The code below shows how `FluxTable` is parsed and returned as a list of dicts
for each record returned.
```python
# iterate through the result(s)
    results = []
    for table in response:
        results.append(table.records[0].values)

    return results
```
```python
[
   {
      'result': '_result', 
       'table': 0, 
       '_start': datetime.datetime(1970, 1, 1, 0, 0, tzinfo=tzutc()), 
       '_stop': datetime.datetime(2022, 5, 8, 22, 25, 10, 111697, tzinfo=tzutc()), 
       '_time': datetime.datetime(2022, 5, 5, 17, 25, 48, 57014, tzinfo=tzutc()), 
       '_value': 33.780799865722656, 
       'HumiditySensor': 'virtual_bme280', 
       'PressureSensor': 'virtual_bme280', 
       'TemperatureSensor': 'virtual_bme280', 
       '_field': 'Lat', 
       '_measurement': 'environment', 
       'device': 'test_device_508472435243'
    }
]
```

## Connect the UI to the API
Create a UI that utilizes the application to query and visualize data from InfluxDB.

> Your IoT Dashboard has four main pages.
> 
> - Query Devices
> - Create Devices
> - Write Data
> - Query Data
> 
> In the `./templates` directory, create the following files:
>  
> - `create.html`: [Creates virtual IoT devices.](https://github.com/influxdata/iot-python-api/blob/main/templates/create.html)
> - `data.html`: [Queries telemetry data.](https://github.com/influxdata/iot-python-api/blob/main/templates/data.html)
> - `devices.html`: [Queries and displays device data.](https://github.com/influxdata/iot-python-api/blob/main/templates/devices.html)
> - `write.html`: [Writes telemetry data.](https://github.com/influxdata/iot-python-api/blob/main/templates/write.html)
> 
> After creating those files are created, update `base.html` and `app.py` to connect the routes.
> `app.py` serves the routes and runs the core logic necessary to populate the UI.
> 
> For example, in the `data` route, the app retrieves `device_id_input` from 
> `data.html` when user input is submitted. It then calls `get_measurements()` 
> with the provided input. After the query runs, `data.html` displays query results.
> 

Reviewer 2: ```suggestion
Your IoT Dashboard has the following four main views:

- Query Devices: queries and displays device information.
- Create Devices: creates virtual IoT devices.
- Write Data: writes telemetry data.
- Query Data: queries telemetry data.

In the `./templates` directory, create the following files:
 
- [`create.html`](https://github.com/influxdata/iot-python-api/blob/main/templates/create.html): provides `Create Devices` view.
- [`data.html`](https://github.com/influxdata/iot-python-api/blob/main/templates/data.html): provides `Query Data` view.
- [`devices.html`](https://github.com/influxdata/iot-python-api/blob/main/templates/devices.html): provides `Query Devices` view.
- [`write.html`](https://github.com/influxdata/iot-python-api/blob/main/templates/write.html): provides `Write Data` view.

 With the view files created, update `base.html` and `app.py` to route server requests to your views.
`app.py` serves the routes and runs the core logic necessary to populate the UI.

For example, in the `data` route, `app.py` retrieves `device_id_input` from 
`data.html` once a user submits input in a `POST` request.
`app.py` then passes the form data to `get_measurements(device_id)`.
Once the query completes, `data.html` displays the results.

```

and interact with your InfluxDB instance. The client needs the following 
information to connect to your InfluxDB instance:

* your InfluxDB [API token](#authorization) with permission to query (_read_) buckets
and create (_write_) authorizations for IoT devices.
* your InfluxDB instance url
* your InfluxDB org ID
* your InfluxDB bucket names

To set up the client configuration, create a `config.ini` in your project's top 
level directory.
In `config.ini`, paste the following:

```ini
[APP]
INFLUX_URL = {{INFLUX_URL}}
INFLUX_TOKEN = {{INFLUX_TOKEN}}
INFLUX_ORG = {{INFLUX_ORG_ID}}
INFLUX_BUCKET = {{INFLUX_BUCKET_FOR_TELEMETRY}}
INFLUX_BUCKET_AUTH = {{INFLUX_BUCKET_FOR_DEVICES}}
```
{{% caption %}}[@influxdata/iot-python-app config.ini](https://github.com/influxdata/iot-python-api/blob/main/config.ini){{% /caption %}}

The following is a sample configuration.
```ini
[APP]
INFLUX_URL = https://us-west-2-2.aws.cloud2.influxdata.com/
INFLUX_TOKEN = 52Pc_ZkJsRh1PKzlwrK8yO6jWSDh6WPAHbfqp-5aROz4zBnY2mvkKws9YoYzksGH3_Xp90rVqo2PRiajTxaUcw==
INFLUX_ORG = bea7ea952287f70d
INFLUX_BUCKET = sly's Bucket
INFLUX_BUCKET_AUTH = devices_auth
```

## Create IoT Virtual Device

> The IoT virtual device generates simulated weather data for you to store in
> InfluxDB. 

Reviewer: ```suggestion
The IoT virtual device generates simulated weather data to store in InfluxDB. 
```

> In the `./api` directory of your project, create a `sensor.py` file.

Reviewer: ```suggestion
In the `./api` directory of your project, create a `sensor.py` file and paste the following:
```

> [//]: # Probably preferable to link the file rather than have the whole file written up()

Reviewer: I'd keep the code here, but to shorten the page, you can wrap it in a `{{% truncate %}}` shortcode. This visually truncates the code block, but lets users expand and collapse it.

    response = http.request('GET', url)
    if not 200 <= response.status <= 299:
        raise Exception(f'[HTTP - {response.status}]: {response.reason}')
    config_fresh = json.loads(response.data.decode('utf-8'))
    return config_fresh

class Sensor:
    def __init__(self):
        self.id = ''
        self.temperature = None
        self.pressure = None
        self.humidity = None
        self.geo = None

    def generate_measurement(self):
        return round(random.uniform(0, 100))

    def geo(self):
        """
        Get GEO location from https://freegeoip.app/json/'.
        :return: Returns a dictionary with `latitude` and `longitude` key.
        """
        try:
            return fetch_json('https://freegeoip.app/json/')
        except Exception:
            return {
                'latitude':  self.generate_measurement(),
                'longitude':  self.generate_measurement(),
            }
```
{{% caption %}}[@influxdata/iot-python-app ./api/sensor.py](https://github.com/influxdata/iot-python-api/blob/main/api/sensor.py){{% /caption %}}

The Sensor object's function `generate_measurement()` will be used to simulate 
weather data.

> ## Develop with the API

Reviewer: Don't need this heading.

```suggestion
```

## Store Virtual IoT Device to InfluxDB

In this section, you will learn how to use the python client library to store
virtual device information in InfluxDB.
Virtual device information will be stored in your first bucket. 
Telemetry data from the device will be stored in the second bucket. 

> (You will learn more about this later on in the guide).

Reviewer: This information about buckets needs to be moved to before the configuration section.

In your `./api` directory, create a `devices.py` file. This file holds the 
core functionality for your app.

> 

Reviewer: Here you show the whole `devices.py` file up front. I think it would be better to walk through adding each piece of the file and explain what each does. Then at the end of this section, show the end result (the entire file).

records generated by the query.
```python
def get_device(device_id) -> {}:
    influxdb_client = InfluxDBClient(url=config.get('APP', 'INFLUX_URL'),
                                     token=config.get('APP', 'INFLUX_TOKEN'),
                                     org=config.get('APP', 'INFLUX_ORG'))

    # Queries must be formatted with single and double quotes correctly
    query_api = QueryApi(influxdb_client)
    device_id = str(device_id)
    device_filter = f'r.deviceId == "{device_id}" and r._field != "token"'
    flux_query = f'from(bucket: "{config.get("APP", "INFLUX_BUCKET_AUTH")}") ' \
                 f'|> range(start: 0) ' \
                 f'|> filter(fn: (r) => r._measurement == "deviceauth" and {device_filter}) ' \
                 f'|> last()'
    devices = {}

    response = query_api.query(flux_query)

   results = []
       for table in response:
           for record in table.records:
               results.append((record.get_field(), record.get_value()))
    return results
```
{{% caption %}}[@influxdata/iot-python-app ./api/devices.py get_device() line 31](https://github.com/influxdata/iot-python-api/blob/main/api/devices.py#L31){{% /caption %}}

> Create a `QueryApi` instance. The `QueryApi` instance queries records from a
> specified bucket using Flux. To generate the Flux query, set the bucket 
> information, range, and query filter. This query will filter on `_measurement` 
> value, `deviceauth`, and searches for any device_id that matches the passed in 
> device_id. Add clause to search for `_field`s that do not contain `token` as a 
> value.

Reviewer: ```suggestion
Create a `QueryApi` instance that queries records from devices bucket using Flux.
Create a query template that:

1. Retrieves the device bucket ID to query from the application configuration.
2. Queries the `deviceauth` measurement.
3. Filters points by a specified device ID.
4. Filters out points with the `token` field.
```

**Note:** `start: 0` does not scale well and can be dangerous. I think we should default to -30d, but also make this configurable.

```python
device_filter = f'r.deviceId == "{device_id}" and r._field != "token"'
flux_query = f'from(bucket: "{config.get("APP", "INFLUX_BUCKET_AUTH")}") ' \
           f'|> range(start: 0) ' \
           f'|> filter(fn: (r) => r._measurement == "deviceauth" and {device_filter}) ' \
           f'|> last()'
```

The client returns a `FluxTable`. Parse the object into a list of tuples.
```python
# Samples results
[('key', 'fake_auth_id_1'), ('key', 'fake_auth_id_2')]

> ```

Reviewer: ~~~suggestion
The client returns a `FluxTable`. Parse the object into a list of tuples.

```python
# Samples results
[('key', 'fake_auth_id_1'), ('key', 'fake_auth_id_2')]
```
~~~

Show the code that actually parses the response into tuples.

        .tag("device", device_id) \
        .tag("TemperatureSensor", "virtual_bme280") \
        .tag("HumiditySensor", "virtual_bme280") \
        .tag("PressureSensor", "virtual_bme280") \
        .field("Temperature", virtual_device.generate_measurement()) \
        .field("Humidity", virtual_device.generate_measurement()) \
        .field("Pressure", virtual_device.generate_measurement()) \
        .field("Lat", coord['latitude']) \
        .field("Lon", coord['latitude']) \
        .time(datetime.utcnow())

    print(f"Writing: {point.to_line_protocol()}")
    client_response = write_api.write(bucket=config.get('APP', 'INFLUX_BUCKET'), record=point)

    # write() returns None on success
    if client_response is None:
        # TODO Maybe also return the data that was written
        return device_id

    # Return None on failure
    return None
```
{{% caption %}}[@influxdata/iot-python-app ./api/devices.py write_measurement() line 84](https://github.com/influxdata/iot-python-api/blob/main/api/devices.py#L84){{% /caption %}}

## Query Telemetry Data
After telemetry data is written into your bucket, query InfluxDB to retrieve the 
telemetry data.  

> Within `devices.py` create a new function called `get_measurements()`. 
> This function will take in a `device_id` and query for simulated weather 
> telemetry data produced by the virtual device. Query on the `_measurement`,
> `environment` and search for all records where `device` matches the `device_id`. 
> Parse the returned `FluxTable` and return each record as a dict containing the 
> data from each record returned.

Reviewer: ```suggestion
Within `devices.py` create a new function called `get_measurements()`. 
This function should:

 1. Take a `device_id` and query for simulated weather telemetry data produced by the virtual device.
 2. Query the `environment` measurement.
 3. Filter by `device_id`. 
4. Parse the returned `FluxTable` and return each record as a [dictionary](https://docs.python.org/tutorial/datastructures.html#dictionaries) containing the 
data from each record returned.
```

    results = []
    for table in response:
        results.append(table.records[0].values)

    return results
```
{{% caption %}}[@influxdata/iot-python-app ./api/devices.py get_measurements() line 116](https://github.com/influxdata/iot-python-api/blob/main/api/devices.py#L116){{% /caption %}}

The code below shows how `FluxTable` is parsed and returned as a list of dicts
for each record returned.
```python
# iterate through the result(s)
    results = []
    for table in response:
        results.append(table.records[0].values)

    return results
```
```python
[
   {
      'result': '_result', 
       'table': 0, 
       '_start': datetime.datetime(1970, 1, 1, 0, 0, tzinfo=tzutc()), 
       '_stop': datetime.datetime(2022, 5, 8, 22, 25, 10, 111697, tzinfo=tzutc()), 
       '_time': datetime.datetime(2022, 5, 5, 17, 25, 48, 57014, tzinfo=tzutc()), 
       '_value': 33.780799865722656, 
       'HumiditySensor': 'virtual_bme280', 
       'PressureSensor': 'virtual_bme280', 
       'TemperatureSensor': 'virtual_bme280', 
       '_field': 'Lat', 
       '_measurement': 'environment', 
       'device': 'test_device_508472435243'
    }
]
```

## Connect the UI to the API

> Now that the core functionality has been implemented, we can now create a UI to perform these requests.
> Your IoT Dashboard will have four main pages.
> * Query Devices
> * Create Devices
> * Write Data
> * Query Data

Reviewer: ```suggestion
Create a UI that utilizes the your application to query and visualize data from InfluxDB.

Your IoT Dashboard has four main pages.

- Query Devices
- Create Devices
- Write Data
- Query Data
```

> In the `./templates` directory, create the following files 
> * `create.html`(link outs to the files. UI code explanation out of scope) (Page to create virutal IoT device)
> * `data.html` (Page to query for telemetry data)
> * `devices.html` (Page to query for device data)
> * `write.html` (Page to write telemetry data)

Reviewer: ```suggestion
In the `./templates` directory, create the following files:
 
- `create.html`: Creates virutal IoT devices.
- `data.html`: Queries telemetry data.
- `devices.html`: Queries and displays device data.
- `write.html`: Writes telemetry data.
```

> After those files are created, update `base.html` and `app.py` to connect the routes.
> Without going into details regarding the UI, `app.py` serves the routes and runs
> the core logic. 

Reviewer: ```suggestion
After creating those files are created, update `base.html` and `app.py` to connect the routes.
`app.py` serves the routes and runs the core logic necessary to populate the UI. 
```

> For example, in the `data` route, `device_id_input` is retrieved from 
> `data.html` when user input is submitted. `get_measurements()` is then called 
> with the provided input. After the query is ran, `data.html` re-renders with the
> query results. 

Reviewer: ```suggestion
For example, in the `data` route, the app retrieves `device_id_input` from 
`data.html` when user input is submitted. It then calls `get_measurements()` 
with the provided input. After the query runs, `data.html` displays query results. 
```

```python
@app.route('/data', methods=['GET', 'POST'])
def data():
    if request.method == 'GET':
        return render_template('data.html', data=None)
    else:
        device_id = request.form.get('device_id_input', None)
        results = devices.get_measurements(device_id)
        return render_template('data.html', data=results)
```
{{% caption %}}[@influxdata/iot-python-app app.py line 83](https://github.com/influxdata/iot-python-api/blob/main/app.py#L83){{% /caption %}}

After these files are created and updated, run the app to view the completed 
IoT Center at `http://localhost:5000`.
```bash
flask run

> ```

Author: Thats a good point, changes to the app will only hot reload if the setting/configuration for it has been set. I'll update the guide to walk the user through that setting during the initial start of the application
